MINOR: Add documentation about KIP-405 remote reads serving just one partition per FetchRequest

stanislavkozlovski opened the pull request:
[As discussed in the mailing
list](https://lists.apache.org/thread/m03mpkm93737kk6d1nd6fbv9wdgsrhv9),
the broker only fetches remote data for ONE partition in a given
FetchRequest. In other words, if a consumer sends a FetchRequest
requesting 50 topic-partitions, and each partition's requested offset is
not stored locally - the broker will fetch and respond with just one
partition's worth of data from the remote store, and the rest will be
empty.

Given our defaults for total fetch response is 50 MiB and per partition
is 1 MiB, this can limit throughput. This patch documents the behavior
in 3 configs - `fetch.max.bytes`, `max.partition.fetch.bytes` and
`remote.fetch.max.wait.ms`

Reviewers: Luke Chen <showuon@gmail.com>, Kamal Chandraprakash
 <kamal.chandraprakash@gmail.com>, Satish Duggana <satishd@apache.org>

stanislavkozlovski committed:
MINOR: Add documentation about KIP-405 remote reads serving just one partition per FetchRequest

showuon reviewed (commented):
LGTM! Do we want to add something in the limitation of tiered storage [here](https://kafka.apache.org/documentation/#tiered_storage_limitation)?

stanislavkozlovski committed:
Fix some issues in docs, add limitation there, also mention an apache licensed pluggin for tiered storage

stanislavkozlovski commented:
I was never able to find the GA release notes :( The docs are so buried.

Added the info to the limitations.

Additionally, I:
1. edited the KIP to mention its GA status and link to the docs - https://cwiki.apache.org/confluence/display/KAFKA/KIP-405%3A+Kafka+Tiered+Storage
2. added the limitation to its release notes page - https://cwiki.apache.org/confluence/display/KAFKA/Kafka+Tiered+Storage+GA+Release+Notes

![image](https://github.com/user-attachments/assets/6f274488-27fa-4d40-b88a-88d20c217f61)
![image](https://github.com/user-attachments/assets/e6f73b0f-1078-4184-bf39-a6cd0a17b6a1)

I also fixed a small bug in the Tiered Storage docs heading

stanislavkozlovski commented on docs/ops.html:
```diff
@@ -4026,13 +4026,14 @@ <h4 class="anchor-heading"><a id="tiered_storage_config" class="anchor-link"></a
-  and <code>remote.log.storage.manager.class.path</code> to specify the implementation of RemoteStorageManager.
+  doesn't provide out-of-the-box implementation of RemoteStorageManager. Users must configure <code>remote.log.storage.manager.class.name</code>
+  and <code>remote.log.storage.manager.class.path</code> to specify the implementation of RemoteStorageManager. One Apache 2.0 licensed implementation supporting all
+  three major hyperscalers' object stores (AWS, GCP, Azure) is <a href="https://github.com/Aiven-Open/tiered-storage-for-apache-kafka">available here</a>.
```
I am not affiliated with Aiven in doing this, but I believe it's of service to the community to link to the only production grade plugin which supports all 3 major clouds.

lmk if you don't believe this is a good idea, i can be convinced to remove it (e.g if we have strict policies, etc). Since it's deep in the docs and is Apache licensed code, I don't see an issue in doing this

showuon replied:
I don't think it's a good idea to mention any 3rd party project in the doc.

showuon replied:
@stanislavkozlovski , I think we should remove anything mentioning the 3rd party project in the doc.

satishd replied:
@stanislavkozlovski  afaik, we do not provide or mention any 3rd party projects in the docs. You can remove it in this PR. We can have a followup PR and discussion on it.

stanislavkozlovski replied:
done

stanislavkozlovski commented on docs/ops.html:
```diff
@@ -4009,7 +4009,7 @@ <h4 class="anchor-heading"><a id="kraft_zk_migration" class="anchor-link"></a><a
     See the <a href="/39/documentation/#kraft_zk_migration">ZooKeeper to KRaft Migration steps</a> in the 3.9 documentation.</p>
 
-<h3 class="anchor-heading"><a id="tiered_storage" class="anchor-link"></a><a href="#kraft">6.9 Tiered Storage</a></h3>
+<h3 class="anchor-heading"><a id="tiered_storage" class="anchor-link"></a><a href="#tiered_storage">6.9 Tiered Storage</a></h3>
```
this was an annoying bug i found

showuon replied:
Nice find!

kamalcph commented on clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerConfig.java:
```diff
@@ -195,7 +195,10 @@ public class ConsumerConfig extends AbstractConfig {
+            "<code>max.message.bytes</code> (topic config). A fetch request consists of many partitions, and there is another setting that controls how much " +
+            "data is returned for each partition in a fetch request - see <code>max.partition.fetch.bytes</code>. Note that there is a current limitation when " +
+            "performing remote reads from tiered storage (KIP-405) - only one partition out of the fetch request is fetched from the remote store (KAFKA-14915). " +
+            "Note also that the consumer performs multiple fetches in parallel.";
```
> Note also that the consumer performs multiple fetches in parallel.

This statement is unclear to me. Does this refer to: 

1. Consumer reads from multiple brokers as the partitions gets spread out?
2. Multiple consumers with different client-id reading from a broker?

stanislavkozlovski replied:
I assume it means a single consumer will send multiple fetch requests. Note @kamalcph I didn't add this - it was here

kamalcph commented on clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerConfig.java:
```diff
@@ -220,7 +223,9 @@ public class ConsumerConfig extends AbstractConfig {
             "accepted by the broker is defined via <code>message.max.bytes</code> (broker config) or " +
-            "<code>max.message.bytes</code> (topic config). See " + FETCH_MAX_BYTES_CONFIG + " for limiting the consumer request size.";
+            "<code>max.message.bytes</code> (topic config). See " + FETCH_MAX_BYTES_CONFIG + " for limiting the consumer request size. " +
+            "Consider increasing this limit especially in the cases of remote storage reads (KIP-405), because currently only " +
```
> Consider increasing this limit

Could this be changed to?
```
Consider increasing the <code>max.partition.fetch.bytes</code> limit ...
```

If the user increase the `max.partition.fetch.bytes` from 1 -> 50 MB to match with the `fetch.max.bytes`, then it may cause GC pressure on the broker due to holding of large continuous byte array.

stanislavkozlovski replied:
good point

kamalcph commented on docs/ops.html:
```diff
@@ -4169,6 +4170,7 @@ <h4 class="anchor-heading"><a id="tiered_storage_limitation" class="anchor-link"
   <li>Deleting tiered storage enabled topics is required before disabling tiered storage at the broker level</li>
```
Please also reword this sentence:

Deleting tiered storage enabled topics  ... -> Disabling tiered storage enabled topics ...

stanislavkozlovski replied:
done

stanislavkozlovski committed:
address comments

stanislavkozlovski commented:
@kamalcph @satishd @showuon can i mereg this?

satishd reviewed (commented):
Thanks @stanislavkozlovski for the PR, overall LGTM. I left a reply/comment.

kamalcph commented:
@stanislavkozlovski 

Could you address the pending comments in the PR? Thanks!

stanislavkozlovski committed:
remove third party lib mention

stanislavkozlovski commented:
should be good to merge @satishd @kamalcph @showuon

showuon reviewed (approved):
LGTM!

kamalcph reviewed (approved):
LGTM!
